libtock/tock.c

```diff
@@ -35,22 +35,102 @@ int tock_enqueue(subscribe_upcall cb, int arg0, int arg1, int arg2, void* ud) {
   return task_last;
 }
 
-int tock_error_to_rcode(tock_error_t err) {
-  switch (err) {
-    case TOCK_ERROR_FAIL:        return TOCK_FAIL;
-    case TOCK_ERROR_BUSY:        return TOCK_EBUSY;
-    case TOCK_ERROR_ALREADY:     return TOCK_EALREADY;
-    case TOCK_ERROR_OFF:         return TOCK_EOFF;
-    case TOCK_ERROR_RESERVE:     return TOCK_ERESERVE;
-    case TOCK_ERROR_INVAL:       return TOCK_EINVAL;
-    case TOCK_ERROR_SIZE:        return TOCK_ESIZE;
-    case TOCK_ERROR_CANCEL:      return TOCK_ECANCEL;
-    case TOCK_ERROR_NOMEM:       return TOCK_ENOMEM;
-    case TOCK_ERROR_NOSUPPORT:   return TOCK_ENOSUPPORT;
-    case TOCK_ERROR_NODEVICE:    return TOCK_ENODEVICE;
-    case TOCK_ERROR_UNINSTALLED: return TOCK_EUNINSTALLED;
-    case TOCK_ERROR_NOACK:       return TOCK_ENOACK;
-    default:                     return TOCK_FAIL;
+int tock_status_to_returncode(statuscode_t status) {
+  // Conversion is easy. Since ReturnCode numeric mappings are -1*ErrorCode,
+  // and success is 0 in both cases, we can just multiply by -1.
+  return -1 * status;
+}
+
+int tock_command_return_novalue_to_returncode(syscall_return_t command_return) {
+  if (command_return.type == TOCK_SYSCALL_SUCCESS) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE) {
+    return tock_status_to_returncode(command_return.data[0]);
+
+    // the remaining cases should not happen if using this function.
+
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U32_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U64) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U32_U32_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_SUCCESS_U64_U32) {
+    return RETURNCODE_SUCCESS;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U32) {
+    return RETURNCODE_FAIL;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U32_U32) {
+    return RETURNCODE_FAIL;
+  } else if (command_return.type == TOCK_SYSCALL_FAILURE_U64) {
+    return RETURNCODE_FAIL;
+  }
```

Comment:
I think all these cases should probably just be replaced by `return RETURNCODE_EBADRVAL` -- if someone calls this function with the wrong comand_return type, it shouldn't "kinda work"